pkg/apis/environment/extension_view.go

```diff
@@ -54,3 +71,125 @@ type (
+				templateversion.FieldTemplateID,
+				templateversion.FieldName,
+				templateversion.FieldVersion,
+				templateversion.FieldProjectID)
+		}).
+		WithProject(func(pq *model.ProjectQuery) {
+			pq.Select(project.FieldName)
+		}).
+		WithEnvironment(func(eq *model.EnvironmentQuery) {
+			eq.Select(environment.FieldName)
+		}).
+		All(r.Context)
+	if err != nil {
+		return err
+	}
+
+	stoppableResources := make(model.Resources, 0, len(resources))
+	stoppableResourceIDs := make([]object.ID, 0, len(resources))
+
+	for _, r := range resources {
+		if pkgresource.IsStoppable(r) {
+			stoppableResources = append(stoppableResources, r)
+			stoppableResourceIDs = append(stoppableResourceIDs, r.ID)
+		}
+	}
+
+	r.stoppableResources = stoppableResources
+
+	dependantIDs, err := dao.GetResourceDependantIDs(r.Context, r.Client, stoppableResourceIDs...)
+	if err != nil {
+		return fmt.Errorf("failed to get resource dependencies: %w", err)
+	}
+
+	dependantIDSet := sets.New[object.ID](dependantIDs...)
+	toStopIDSet := sets.New[object.ID](stoppableResourceIDs...)
+
+	// Validate if a resource is about to stop but the resources depending on it is not.
+	diffIDSet := dependantIDSet.Difference(toStopIDSet)
+	if diffIDSet.Len() > 0 {
+		names, err := dao.GetResourceNamesByIDs(r.Context, r.Client, diffIDSet.UnsortedList()...)
+		if err != nil {
+			return fmt.Errorf("failed to get resources: %w", err)
+		}
+
+		return errorx.HttpErrorf(
+			http.StatusConflict,
+			"resource about to be stopped is the dependency of: %v",
+			strs.Join(", ", names...),
+		)
+	}
+
+	return nil
+}
+
+type RouteStartRequest struct {
+	_ struct{} `route:"POST=/start"`
+
+	model.EnvironmentQueryInput `path:",inline"`
+
+	stoppableResources model.Resources `json:"-"`
+}
+
+func (r *RouteStartRequest) Validate() error {
+	if err := r.EnvironmentQueryInput.Validate(); err != nil {
+		return err
+	}
+
+	// Check revision status of resources within the environment.
+	resources, err := r.Client.Resources().Query().
+		Where(resource.EnvironmentID(r.ID)).
+		WithTemplate(func(tvq *model.TemplateVersionQuery) {
+			tvq.Select(
+				templateversion.FieldID,
+				templateversion.FieldTemplateID,
+				templateversion.FieldName,
+				templateversion.FieldVersion,
+				templateversion.FieldProjectID)
+		}).
+		WithProject(func(pq *model.ProjectQuery) {
+			pq.Select(project.FieldName)
+		}).
+		WithEnvironment(func(eq *model.EnvironmentQuery) {
+			eq.Select(environment.FieldName)
+		}).
+		All(r.Context)
+	if err != nil {
+		return err
+	}
+
+	stoppableResources := make(model.Resources, 0, len(resources))
+	stoppableResourceIDs := make([]object.ID, 0, len(resources))
+
+	for _, r := range resources {
+		if pkgresource.IsStoppable(r) {
+			stoppableResources = append(stoppableResources, r)
+			stoppableResourceIDs = append(stoppableResourceIDs, r.ID)
+		}
+	}
+
+	r.stoppableResources = stoppableResources
+
+	return apiresource.ValidateRevisionsStatus(r.Context, r.Client, stoppableResourceIDs...)
```

Comment:
What is the difference between start and upgrade when resource is Ready or Failed in this request?

Reply:
Update validation so that start only applies to undeployed/stopped resources.